src/darwin/Framework/CHIP/MTRSetupPayload.mm

```diff
@@ -19,251 +19,503 @@
+- (instancetype)initWithTag:(uint8_t)tag stringValue:(NSString *)value
 {
-    if (!value.HasValue()) {
-        return MTRDiscoveryCapabilitiesUnknown;
-    }
+    self = [super init];
+    _info.type = chip::optionalQRCodeInfoTypeString;
+    _info.tag = tag;
+    _info.data = value.UTF8String;
+    return self;
+}
 
-    NSUInteger flags = 0;
-    if (value.Value().Has(chip::RendezvousInformationFlag::kBLE)) {
-        flags |= MTRDiscoveryCapabilitiesBLE;
-    }
-    if (value.Value().Has(chip::RendezvousInformationFlag::kSoftAP)) {
-        flags |= MTRDiscoveryCapabilitiesSoftAP;
-    }
-    if (value.Value().Has(chip::RendezvousInformationFlag::kOnNetwork)) {
-        flags |= MTRDiscoveryCapabilitiesOnNetwork;
-    }
-    if (flags == MTRDiscoveryCapabilitiesUnknown) {
-        // OnNetwork has to be supported!
-        flags = MTRDiscoveryCapabilitiesOnNetwork;
-    }
-    return flags;
+- (nullable instancetype)initWithQRCodeInfo:(chip::OptionalQRCodeInfo const &)info
+{
+    self = [super init];
+    _info = info;
+    // Don't expose objects with an invalid type.
+    VerifyOrReturnValue(self.type != MTROptionalQRCodeInfoTypeUnknown, nil);
+    return self;
 }
 
-+ (chip::Optional<chip::RendezvousInformationFlags>)convertDiscoveryCapabilities:(MTRDiscoveryCapabilities)value
+- (CHIP_ERROR)addAsVendorElementTo:(chip::SetupPayload &)payload
 {
-    if (value == MTRDiscoveryCapabilitiesUnknown) {
-        return chip::NullOptional;
+    switch (_info.type) {
+    case chip::optionalQRCodeInfoTypeString:
+        return payload.addOptionalVendorData(_info.tag, _info.data);
+    case chip::optionalQRCodeInfoTypeInt32:
+        return payload.addOptionalVendorData(_info.tag, _info.int32);
+    default:
+        return CHIP_ERROR_INCORRECT_STATE;
     }
+}
 
-    chip::RendezvousInformationFlags flags;
-    if (value & MTRDiscoveryCapabilitiesBLE) {
-        flags.Set(chip::RendezvousInformationFlag::kBLE);
-    }
+- (MTROptionalQRCodeInfoType)type
+{
+    // The way OptionalQRCodeInfo uses what are effectively C type identifiers (uint32 etc)
+    // rather than TLV types is not ideal. If we add support for additional integer types
+    // we should consider replacing MTROptionalQRCodeInfoTypeInt32 with
+    // MTROptionalQRCodeInfoTypeInteger and hiding the low-level C representation.
+    switch (_info.type) {
+    case chip::optionalQRCodeInfoTypeString:
+        return MTROptionalQRCodeInfoTypeString;
+    case chip::optionalQRCodeInfoTypeInt32:
+        return MTROptionalQRCodeInfoTypeInt32;
+    // No 'default:' so we get a warning if new types are added.
+    // OptionalQRCodeInfo does not support these types
+    case chip::optionalQRCodeInfoTypeInt64:
+    case chip::optionalQRCodeInfoTypeUInt32:
+    case chip::optionalQRCodeInfoTypeUInt64:
+    // We should never see the unknown type
+    case chip::optionalQRCodeInfoTypeUnknown:
+        /* fall through */;
```

Comment:
`[[fallthrough]]`?

Reply:
Turns out `[[fallthrough]]` is not valid at the end of the switch statement (must have another case or default after it)